rest_framework/permissions.py

```diff
@@ -71,27 +73,35 @@ def __init__(self, op1, op2):
         self.op2 = op2
 
     def has_permission(self, request, view):
-        return (
-            self.op1.has_permission(request, view) or
-            self.op2.has_permission(request, view)
-        )
+        hasperm1 = self.op1.has_permission(request, view)
+        if hasperm1 and hasperm1 is not NotImplemented:
+            return hasperm1
+        hasperm2 = self.op2.has_permission(request, view)
+        return hasperm2 or hasperm1
 
     def has_object_permission(self, request, view, obj):
-        return (
-            self.op1.has_object_permission(request, view, obj) or
-            self.op2.has_object_permission(request, view, obj)
-        )
+        hasperm1 = self.op1.has_object_permission(request, view, obj)
+        if hasperm1 is NotImplemented:
+            hasperm1 = self.op1.has_permission(request, view)
```

Comment:
What would happen if these lines are removed?

Reply:
TL;DR If I wrote my tests right, if you remove these two lines and run the tests, you'll see exactly why they are needed.

Long version: If `has_object_permission` returns `NotImplemented` it should fall back on `has_permission`, but this is only necessary in the OR case. With a non-composed permission or an AND case, calling `has_permission` is redundant because you know it must have returned True in the original `has_permission` check to make it to check the object. But with OR, it's possible that `has_permission` is False and the original check passed only because of the other side of the OR. We don't want the `has_object_permission` test to pass for the wrong reason.